docs/patterns/outbox.md

```diff
@@ -0,0 +1,141 @@
+
+The assumption is that the _side_ effect application never fails. The resulting algorithm for the Outbox pattern is following:
+
+```c#
+if (HasBeenProcessed(message))
+{
+    return;
+}
+ComputeEffects(message);
+var tx = BeginTransaction();
+ApplyEffectsToPrimary(tx);
+StoreSideEffects(tx);
+MarkProcessed(message, tx);
+tx.Commit();
+
+ApplySideEffects();
+CleanUpSideEffects();
+```
+
+The Outbox pattern can be implemented in the infrastructure layer of the solution ensuring that all message handlers running on top of that infrastructure are automatically made _idempotent_. The alternative is ensure each individual message handler is _idempotent_ by applying techniques some of the following techniques:
+
+- do not access more than one resource in any message handler
+- use deterministic or domain identifiers instead of GUIDs
+- do not rely on system clock or pseudo-random value generators
+- use idempotent-by-design data structures
+
+### Benefits
+
+- Provides a simple programming abstraction, similar to one exposed by distributed transaction technologies
+- Can be implemented in the infrastructure layer with minimal effect to message handling code, allowing smooth transition away from distributed transaction technologies
+- Supports resources that never had been able to participate in distributed transactions, such as key-value stores
+- Has relatively mild requirements for the primary resource -- optimistic concurrency control
+
+### Drawbacks
+
+- Generates more traffic with the primary resource (database) in form of two additional operations requiring a round-trip: checking if a message has been processed and cleaning up the side effects.
+- Requires that the primary resource supports optimistic concurrency control
+- Requires storing information about all previously processed messages. In principle that information should never be removed but in practice often age-based approach is used to prevent data from growing indefinitely. 
+
+## Issues and considerations
+
+Consider the following points when implementing the Outbox pattern:
+
+- In order to take advantage of the pattern, each message needs to have a unique identifier. Many messaging technologies like Azure Service Bus provide this identifier out-of-the-box, but in some of them, it is not set automatically.
+- Consider how long to store information about processed messages. The likelihood of message being duplicated diminishes with time significantly. In practice value of 1 or 2 weeks is often used as maximum age to keep the deduplication data.
+- Depending on the technology used for the primary resource, it might or might not be possible to use native features to remove expired dedupliction data. For example, Cosmos DB has a built in time-to-live mechanism that can be used to ensure message processing information is removed after configured time.
```

Comment:
```suggestion
- Depending on the technology used for the primary resource, it might or might not be possible to use native features to remove expired dedupliction data. Cosmos DB users can take advantage of the built-in time-to-live mechanism to ensure message processing information is removed after the configured time.
```